Batteries/Data/Vector/Lemmas.lean

```diff
@@ -0,0 +1,59 @@
+import Batteries.Data.Vector.Basic
+import Batteries.Data.List.Basic
+import Batteries.Data.List.Lemmas
+import Batteries.CodeAction
+
+namespace Vector
+
+
+/-- A `nil` vector maps to a `nil` vector. -/
+@[simp]
+theorem map_nil (f : α → β) : map f empty = empty :=
+  rfl
+
+
+theorem eq : ∀ v w: Vector α n,
+  v.toArray = w.toArray → v = w := by
+  intro v w h
+  cases v with
+  | mk v_arr vh =>
+    cases w with
+    | mk w_arr wh =>
+      simp only at h
+      rw [h] at vh
+      simp only [h]
+
+/-- A vector of length `0` is a `nil` vector. -/
+protected theorem eq_nil (v : Vector α 0) : v = empty := by
+  apply Vector.eq v #v[]
+  apply Array.eq_empty_of_size_eq_zero v.2
+  done
+
+theorem extAux (a b : Vector α n)
+    (h : (i : Nat) → (hi₁ : i < a.toArray.size)
+    → (hi₂ : i < b.toArray.size)
+    → a.toArray[i] = b.toArray[i])
+    : a = b := by
+    have h := Array.ext
+      a.toArray
+      b.toArray
+      (Eq.trans a.size_eq b.size_eq.symm)
+      h
+    apply eq
+    exact h
+    done
+
+@[ext]
+theorem ext (a b :Vector α n)
+  (h : (i : Nat) → (h : i < n) → get a ⟨i,h⟩ = get b ⟨i,h⟩) : a = b := by
+  apply extAux
+  intro i hi₁ hi₂
+  rw [a.size_eq] at hi₁
+  replace h := h i hi₁
+  simp [get] at h
+  exact h
+  done
```

Comment:
```suggestion
@[ext] 
protected theorem ext (a b : Vector α n) (h : (i : Nat) → (_ : i < n) → a[i] = b[i]) : a = b := by
  apply Vector.eq
  apply Array.ext
  · rw [a.size_eq, b.size_eq]
  · intro i hi _
    rw [a.size_eq] at hi
    exact h i hi
```

Reply:
Done. Removed it. Gosh  why didn't I think of that. It is so much better.